File: pkg/runtime/runtimes/go/tests/types_test.go
package main

import (
	"testing"

	"pulserpc-go-runtime/pulserpc"
)

func TestFindStruct(t *testing.T) {
	allStructs := pulserpc.StructMap{
		"TestStruct": pulserpc.StructDef{
			"fields": []interface{}{
				map[string]interface{}{
					"name": "field1",
					"type": map[string]interface{}{"builtIn": "string"},
				},
			},
		},
	}

	structDef := pulserpc.FindStruct("TestStruct", allStructs)
	if structDef == nil {
		t.Error("Expected to find TestStruct")
	}

	notFound := pulserpc.FindStruct("NotFound", allStructs)
	if notFound != nil {
		t.Error("Expected NotFound to return nil")
	}
}

func TestFindEnum(t *testing.T) {
	allEnums := pulserpc.EnumMap{
		"TestEnum": pulserpc.EnumDef{
			"values": []interface{}{
				map[string]interface{}{"name": "VALUE1"},
				map[string]interface{}{"name": "VALUE2"},
			},
		},
	}

	enumDef := pulserpc.FindEnum("TestEnum", allEnums)
	if enumDef == nil {
		t.Error("Expected to find TestEnum")
	}

	// Test qualified name
	enumDef2 := pulserpc.FindEnum("ns.TestEnum", allEnums)
	if enumDef2 == nil {
		t.Error("Expected to find TestEnum via qualified name")
	}

	notFound := pulserpc.FindEnum("NotFound", allEnums)
	if notFound != nil {
		t.Error("Expected NotFound to return nil")
	}
}

func TestGetStructFields(t *testing.T) {
	allStructs := pulserpc.StructMap{
		"Parent": pulserpc.StructDef{
			"fields": []interface{}{
				map[string]interface{}{
					"name": "parentField",
					"type": map[string]interface{}{"builtIn": "string"},
				},
			},
		},
		"Child": pulserpc.StructDef{
			"extends": "Parent",
			"fields": []interface{}{
				map[string]interface{}{
					"name": "childField",
					"type": map[string]interface{}{"builtIn": "int"},
				},
			},
		},
	}

	fields := pulserpc.GetStructFields("Child", allStructs)
	if len(fields) != 2 {
		t.Errorf("Expected 2 fields (parent + child), got %d", len(fields))
	}

	// Check parent field
	if fields[0]["name"] != "parentField" && fields[1]["name"] != "parentField" {
		t.Error("Expected to find parentField in fields")
	}

	// Check child field
	if fields[0]["name"] != "childField" && fields[1]["name"] != "childField" {
		t.Error("Expected to find childField in fields")
	}
}

File: pkg/runtime/runtimes/java/tests/TypesTest.java
import barrister2.*;
import org.junit.Test;
import org.junit.Assert;
import java.util.*;

public class TypesTest {

    @Test
    public void testFindStruct() {
        Map<String, Map<String, Object>> allStructs = new HashMap<>();

        Map<String, Object> personStruct = new HashMap<>();
        personStruct.put("fields", Arrays.asList(
            Map.of("name", "name", "type", Map.of("builtIn", "string")),
            Map.of("name", "age", "type", Map.of("builtIn", "int"))
        ));
        allStructs.put("Person", personStruct);

        // Found
        Map<String, Object> result = Types.findStruct("Person", allStructs);
        Assert.assertNotNull(result);
        Assert.assertEquals(personStruct, result);

        // Not found
        Map<String, Object> notFound = Types.findStruct("NonExistent", allStructs);
        Assert.assertNull(notFound);
    }

    @Test
    public void testFindEnum() {
        Map<String, Map<String, Object>> allEnums = new HashMap<>();

        Map<String, Object> colorEnum = new HashMap<>();
        colorEnum.put("values", Arrays.asList(
            Map.of("name", "RED"),
            Map.of("name", "GREEN"),
            Map.of("name", "BLUE")
        ));
        allEnums.put("Color", colorEnum);

        // Found
        Map<String, Object> result = Types.findEnum("Color", allEnums);
        Assert.assertNotNull(result);
        Assert.assertEquals(colorEnum, result);

        // Not found
        Map<String, Object> notFound = Types.findEnum("NonExistent", allEnums);
        Assert.assertNull(notFound);
    }

    @Test
    public void testGetStructFields() {
        Map<String, Map<String, Object>> allStructs = new HashMap<>();

        // Base struct
        Map<String, Object> baseStruct = new HashMap<>();
        baseStruct.put("fields", Arrays.asList(
            Map.of("name", "id", "type", Map.of("builtIn", "int"))
        ));
        allStructs.put("Base", baseStruct);

        // Derived struct
        Map<String, Object> derivedStruct = new HashMap<>();
        derivedStruct.put("extends", "Base");
        derivedStruct.put("fields", Arrays.asList(
            Map.of("name", "name", "type", Map.of("builtIn", "string"))
        ));
        allStructs.put("Person", derivedStruct);

        List<Map<String, Object>> fields = Types.getStructFields("Person", allStructs);

        // Should have fields from both base and derived class
        Assert.assertEquals(2, fields.size());

        // Check field names (order may vary due to inheritance resolution)
        Set<String> fieldNames = new HashSet<>();
        for (Map<String, Object> field : fields) {
            fieldNames.add((String) field.get("name"));
        }
        Assert.assertTrue(fieldNames.contains("id"));
        Assert.assertTrue(fieldNames.contains("name"));
    }
}
